### PR TITLE
[vividus] Disalow empty tables for cartesian product transformer

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java [hunk 1]
```diff
@@ -47,19 +55,78 @@
...
+                                           checkEmptyTables(wps);
+                                       }
+
+                                       return wps.stream()
+                                                 .map(DescriptiveTableWrapper::getTable)
+                                                 .collect(Collectors.toCollection(LinkedList::new));
+                                   }));
+    }
+
+    private void checkEmptyTables(List<DescriptiveTableWrapper> wps)
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java:89]: `wps ` ?

File: vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java [hunk 2]
```diff
@@ -47,19 +55,78 @@
...
+    {
+        List<String> emptyTables = wps.stream()
+                                      .filter(w -> w.getTable().getRowCount() == 0)
+                                      .map(DescriptiveTableWrapper::getDescription)
+                                      .collect(Collectors.toList());
+
+        isTrue(emptyTables.isEmpty(), "Empty examples tables were found: %s", emptyTables);
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java:96]: `Empty ExamplesTable-s are not allowed, but %s is/are empty`

File: vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java [hunk 3]
```diff
@@ -47,19 +55,78 @@
...
-            examplesTablesStream = Stream.concat(examplesTablesStream, Stream.of(tableAsString));
+            examplesTablesStream = Stream.concat(examplesTablesStream,
+                    Stream.of(new DescriptiveTableWrapper("input table", tableAsString)));
         }
 
         ExamplesTableFactory factory = configuration.examplesTableFactory();
-        return examplesTablesStream.map(factory::createExamplesTable)
-                                   .collect(Collectors.toCollection(LinkedList::new));
+
+        return examplesTablesStream.map(w -> w.withTable(factory.createExamplesTable(w.getRaw())))
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java:75]: overhead, simple `for` will work better here

File: docs/modules/commons/pages/table-transformers.adoc [hunk 4]
```diff
@@ -309,6 +309,7 @@ Examples:
 ====
 * headers of input tables must be different
 * number of input tables must be more than 1
+* empty tables are not allowed
```
Comment (valfirst) [inline docs/modules/commons/pages/table-transformers.adoc:312]: I guess similar, but opposite message should be added to `MERGING` transformer

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java [hunk 5]
```diff
@@ -36,6 +37,13 @@
 {
     @Inject private Configuration configuration;
 
+    private final boolean allowEmptyTables;
+
+    public AbstractTableLoadingTransformer(boolean allowEmptyTables)
```
Comment (ikalinin1) [inline vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java:42]: protected ?

File: vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java [hunk 6]
```diff
@@ -47,19 +55,78 @@
                 .filter(not(String::isBlank))
                 .collect(Collectors.toList());
 
-        Stream<String> examplesTablesStream = tables.stream();
+        Stream<DescriptiveTableWrapper> examplesTablesStream = IntStream.range(0, tables.size())
```
Comment (ikalinin1) [inline vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java:58]: Not necessary when allowEmptyTables = true ?

Comment (uarlouski) [reply vividus/src/main/java/org/vividus/bdd/transformer/AbstractTableLoadingTransformer.java:60]: Imho its reasonable price to keep the code straightforward

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**